In [10]:
#Import des bibliothèques
import numpy as np
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

In [2]:
#Chargement du dataset
breast_cancer=load_breast_cancer()
X=breast_cancer.data
Y=breast_cancer.target
print("Taille de X:",X.shape)
print("Taille de Y:",Y.shape)
feature_names=breast_cancer.feature_names
class_names=breast_cancer.target_names
print("Les variables explicatives:",feature_names)
print("Les noms des classes:",class_names)
import pandas as pd
#pour visualiser les premieres lignes on transforme le dataset en DataFrame
df=pd.DataFrame(X,columns=feature_names)
print(df.head())

Taille de X: (569, 30)
Taille de Y: (569,)
Les variables explicatives: ['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']
Les noms des classes: ['malignant' 'benign']
   mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20

In [3]:
#Separation de données train et test
(X_train,X_test,Y_train,Y_test)=train_test_split(X,Y,test_size=0.3,random_state=42)

In [21]:
# Mise à l'échelle des données
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Entrainer le modèle LogisticRegression avec les données mises à l'échelle
model_scaled = LogisticRegression(max_iter=1000) #j'ai ajouté max_iter=1000 pour que Le modèle prend le temps de converger mathématiquement vers la meilleure solution possible.
model_scaled.fit(X_train_scaled, Y_train);

## Pourquoi on a standarisé les données pour LogisticRegression:
la standardisation des données n’est pas une étape optionnelle mais une nécessité pour exploiter pleinement le potentiel de la régression logistique. 
En ramenant toutes les variables à une moyenne de 0 et un écart-type de 1, elle résout trois défis majeurs : 
- Elle accélère drastiquement la convergence de la descente de gradient
- garantit une application juste et efficace de la régularisation ($L_1$/$L_2$)
- Rend les coefficients finaux directement comparables et interprétables.
C'est l'étape fondamentale qui transforme un modèle instable en un classifieur performant, robuste et explicable.

In [22]:
#Prédiction
Y_pred=model_scaled.predict(X_test_scaled)

In [23]:
#Evaluation
accuracy=accuracy_score(Y_test,Y_pred)
print("Accuracy du modele Logistic Regression:",accuracy)
print("Matrice de confusion:",confusion_matrix(Y_test,Y_pred))
print("Rapport de classification:")
print(classification_report(Y_test,Y_pred))


Accuracy du modele Logistic Regression: 0.9824561403508771
Matrice de confusion: [[ 62   1]
 [  2 106]]
Rapport de classification:
              precision    recall  f1-score   support

           0       0.97      0.98      0.98        63
           1       0.99      0.98      0.99       108

    accuracy                           0.98       171
   macro avg       0.98      0.98      0.98       171
weighted avg       0.98      0.98      0.98       171



# Conclusion
Le modèle de Régression Logistique affiche d'excellentes performances avec une exactitude globale (Accuracy) de 98,25 %. L'ajustement des hyperparamètres notamment l'augmentation du nombre maximal d'itérations (max_iter=1000) a permis de stabiliser la convergence du modèle. 

La matrice de confusion et le rapport de classification confirment la robustesse du classifieur, qui obtient des scores de précision et de rappel (recall) équilibrés et supérieurs à 97 % pour les deux classes (0 et 1). 

Le modèle démontre ainsi une excellente capacité de généralisation sans signe de biais envers l'une des classes.